src/Java.Interop.Tools.JavaTypeSystem/Adapters/JavaXmlApiExporter.cs

```diff
@@ -0,0 +1,359 @@
+				field.Type,
+				field.TypeGeneric,
+				value,
+				XmlConvert.ToString (field.IsVolatile),
+				null,
+				null,
+				null,
+				null,
+				field.IsNotNull);
+		}
+
+		static void SaveMethod (JavaMethodModel method, XmlWriter writer)
+		{
+			bool check (JavaMethodModel _) => _.BaseMethod?.ParentType?.Visibility == "public" &&
+			     !method.IsStatic &&
+					      method.Parameters.All (p => p.InstantiatedGenericArgumentName == null);
+
+			//// skip synthetic methods, that's what jar2xml does.
+			//// However, jar2xml is based on Java reflection and it generates synthetic methods
+			//// that actually needs to be generated in the output XML (they are not marked as
+			//// "synthetic" either by asm or java reflection), when:
+			//// - the synthetic method is actually from non-public ancestor class
+			////   (e.g. FileBackupHelperBase.writeNewStateDescription())
+			//// For such case, it does not skip generation.
+			if (method.IsSynthetic && (method.BaseMethod == null || check (method)))
+				return;
+
+			//// Here we skip most of the overriding methods of a virtual method, unless
+			//// - the method visibility or final-ity has changed: protected Object#clone() is often
+			////   overriden as public. In that case, we need a "new" method.
+			//// - the method is covariant. In that case we need another overload.
+			//// - they differ in "abstract" or "final" method attribute.
+			//// - the derived method is static.
+			//// - the base method is in the NON-public class.
+			//// - none of the arguments are type parameters.
+			//// - finally, it is the synthetic method already checked above.
+			if (method.BaseMethod != null &&
+			    !method.BaseMethod.IsAbstract &&
+			    method.BaseMethod.Visibility == method.Visibility &&
+			    method.BaseMethod.IsAbstract == method.IsAbstract &&
+			    method.BaseMethod.IsFinal == method.IsFinal &&
+			    !method.IsSynthetic &&
+			    check (method))
+				return;
+
+			SaveMember (method, writer, "method",
+				XmlConvert.ToString (method.IsAbstract),
+				XmlConvert.ToString (method.IsNative),
+				GetVisibleReturnTypeString (method),
+				XmlConvert.ToString (method.IsSynchronized),
+				null,
+				null,
+				null,
+				null,
+				null,
+				method.Parameters,
+				method.IsBridge,
+				method.ReturnJni,
+				method.IsSynthetic,
+				method.ReturnNotNull);
+		}
+
+		static string GetVisibleReturnTypeString (JavaMethodModel method)
+		{
+			if (GetVisibleNonSpecialType (method, method.ReturnTypeModel) is JavaTypeReference jtr)
+				return jtr.ToString ();
+
+			return method.Return;
+		}
+
+		public static string? GetVisibleParamterTypeName (this JavaParameterModel parameter)
+		{
+			if (GetVisibleNonSpecialType (parameter.ParentMethod, parameter.TypeModel) is JavaTypeReference jtr)
+				return jtr.ToString ();
+
+			return parameter.GenericType;
+		}
+
+		static JavaTypeReference? GetVisibleNonSpecialType (JavaMethodModel method, JavaTypeReference? r)
```

Comment:
Naming: I don't understand what a "non-special" type is. It apparently relates to `JavaTypeReference.SpecialName`, but I don't currently understand what *that* means either.

Reply:
c/p from https://github.com/xamarin/java.interop/blob/a3de91efdaec04b3927b755a07018e8d26cfdc71/src/Xamarin.Android.Tools.ApiXmlAdjuster/JavaApiFixVisibilityExtensions.cs#L30-L42 for compatibility